### ISSUE TITLE
Doesn't work with Yarn 2

### ISSUE THREAD

**Issue by schickling** [prisma/prisma2#1439]
When trying to run `prisma2 init` via `yarn dlx` it fails in the following way:

![image](https://user-images.githubusercontent.com/1567498/73133616-a5f09400-402b-11ea-9857-839d3e173986.png)

I'm using yarn `2.0.0-rc.27`

**Comment by divyenduz** [comment]
Internal note: dropping from current sprint

**Comment by yovanoc** [comment]
When have you planned to handle this? Thanks

**Comment by nnennajohn** [comment]
Any update on this? We still have to keep using npx prisma generate, and it creates a node_modules folders which yarn always removes.

**Comment by jgeschwendt** [comment]
I have a simple repro with yarn2 working as best I could get it without modifying the client's output. Hopefully, it can be of some help.

https://github.com/jgeschwendt/prisma-yarn2-repro
https://github.com/jgeschwendt/prisma-yarn2-repro/blob/master/REPRO.md
https://github.com/jgeschwendt/prisma-yarn2-repro/blob/master/yarn-output.png

**Comment by janpio** [comment]
Reproduction with current version of Prisma (2.3.0) and Yarn 2 (2.1):
https://github.com/prisma/prisma/issues/3168#issue-668010124

**Comment by adarnon** [comment]
Is there a plan/roadmap to solving this?

**Comment by janpio** [comment]
Related: https://github.com/prisma/prisma-client-js/issues/819

**Comment by nnennajohn** [comment]
It's been several months since this was opened. Same with an issue with @nexus-prisma-plugin. Sadly, I had to find a different orm solution for now.

**Comment by matthewmueller** [comment]
**Quick update** 

We'd like to add support for Yarn 2, but not just yet. Yarn 2 is a significant departure from other Node package managers (no node_modules!). By default, Prisma Client makes use of node_modules, so we need to figure out how to first detect Yarn 2 and then apply special treatment. 

This is all totally doable, but it will just take time. Right now we're focused on adding more powerful query features to the Prisma Client.

In the meantime, you can workaround this issue by providing a custom output path.

**prisma/schema.prisma**

```prisma
generator client {
  provider = "prisma-client-js"
  // Output to a custom location
  output   = "../generated/client"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model User {
  id      Int      @default(autoincrement()) @id
  email   String   @unique
  name    String?
  posts   Post[]
  profile Profile?
}
```

**main.ts**

```ts
// Include the client in a custom location
import { PrismaClient } from './generated/client'

async function main() {
  const client = new PrismaClient()
  const data = await client.user.findMany()
  console.log(data)
  client.disconnect()
}

main().catch(console.error)
```

You can read more about this feature here: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/generating-prisma-client#specifying-a-custom-output-path

**Comment by TheGrimSilence** [comment]
> We'd like to add support for Yarn 2, but not right now. Yarn 2 is a significant departure from other Node package managers (no node_modules!). By default, Prisma Client relies on the presence of node_modules, so we need to figure out how to first detect Yarn 2 and then apply a Yarn 2-specific solution.

Move `node_modules/.prisma/` -> `<projectRoot>/.prisma/` (universal fix).
If `.yarn/`, Yarn 2 exists.

Yarn 2 already supports TypeScript by adding it's own TypeScript server alongside Pnpify which adds support for various editors like Vim, Visual Studio Code. The `.yarn/*` directory is only generated by Yarn 2, but if you really need a deeper check, `.yarn/cache` and `.yarn/releases` are verifiably *only* Yarn >= 2.0.

Moving `.prisma/` out of node_modules also allows for future compatibility with any other package managers as well.

**Comment by merceyz** [comment]
> If .yarn/, Yarn 2 exists.

That's not necessarily true, the correct check is to see if `yarn.lock` contains `__metadata`, it it does it's Yarn 2

**Comment by TheGrimSilence** [comment]
My bad, thanks @merceyz

**Comment by jgeschwendt** [comment]
I believe you can check the name of the yarn config files to derive the version.

`.yarnrc`, or its absence, is used for v1, `.yarnrc.yml` is used for v2.

https://github.com/yarnpkg/yarn/blob/4df89bc4923a1d24e271d67843cfc19d7fed204a/src/cli/commands/policies.js#L166-L179

Although if you already have detection for the yarn binary checking its range may be better.

https://github.com/yarnpkg/yarn/blob/4df89bc4923a1d24e271d67843cfc19d7fed204a/src/cli/commands/policies.js#L122

**Comment by TheGrimSilence** [comment]
@merceyz would probably have the best input of us since he's a direct contributor to Yarn itself. but I believe you're right.

**Comment by merceyz** [comment]
> .yarnrc, or its absence, is used for v1, .yarnrc.yml is used for v2.

While that is true you can have `.yarnrc.yml` point to a V1 binary to get anyone using V2 globally to use V1

**Comment by zachasme** [comment]
A simpler solution would be a new generator option to set the client package name in addition to output directory.

Right now we output to `./packages/prisma` and replace `"name": ".prisma/client"` (in `./packages/prisma/package.json`) with `@projectname/prisma` (since `.prisma/client` is an invalid package name). Then we add that package as a workspace and use it as a dependency in the other workspaces.

That means we can do
```js
import prisma from "@projectname/prisma"
```

And have prisma working with yarn v2 and workspaces. Only trouble is the invalid package name which means we need to manually edit `package.json` after each client generation.

**Comment by rizkyramadhan** [comment]
https://github.com/prisma/prisma/issues/1439#issuecomment-671861692

This failed for me on windows 10, the error:

  Query engine binary for current platform "windows" could not be found.
This probably happens, because you built Prisma Client on a different platform.
(Prisma Client looked in "\query-engine-windows.exe")

Files in \:

  $Recycle.Bin
  $WinREAgent
  $WINRE_BACKUP_PARTITION.MARKER
  android
  Config.Msi

**Comment by rizkyramadhan** [comment]
> [#1439 (comment)](https://github.com/prisma/prisma/issues/1439#issuecomment-671861692)
> 
> This failed for me on windows 10, the error:
> 
> ```
> Query engine binary for current platform "windows" could not be found.
> This probably happens, because you built Prisma Client on a different platform.
> (Prisma Client looked in "\query-engine-windows.exe")
> 
> Files in \:
> 
> $Recycle.Bin
> $WinREAgent
> $WINRE_BACKUP_PARTITION.MARKER
> android
> Config.Msi
> ```

Solved by setting this on my .env:

```
PRISMA_QUERY_ENGINE_BINARY=pages/api/generated/prisma/query-engine-windows.exe
```

After that , this error happened:

```
  Query engine exited with code 101

thread 'main' panicked at 'Could not open datamodel file "\\schema.prisma": Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', query-engine\query-engine\src\opt.rs:186:17

```

**Comment by samrith-s** [comment]
Does anybody know how to solve for this?

**Comment by TheGrimSilence** [comment]
@samrith-s `yarn config set nodeLinker node-modules && yarn` this will tell Yarn 2 to use `node_modules` which turns of Plug'n'Play mode. It's a quick fix.

**Comment by samrith-s** [comment]
@TheGrimSilence it actually works by simply setting the output directory within `src` and adding that to `.gitignore` and importing like so:
```ts
import { PrismaClient } from '@prisma-client';
```
And have `paths` setup in `tsconfig.json`

**Comment by TheGrimSilence** [comment]
@samrith-s That's *also* an option 😅

**Comment by timsuchanek** [comment]
I just created a separate issue to support Yarn 2 without PnP enabled: https://github.com/prisma/prisma/issues/4493

**Comment by simnalamburt** [comment]
Here's how to make "@prisma/client" work properly with Yarn 2 PnP.

### Instruction
1.  Make ".prisma" as a new dependency of "@prisma/client" package.

    ```yml
    # .yarnrc.yml
    packageExtensions:
	  '@prisma/client@*':
	    dependencies:
	      '.prisma': 'link:See "resolution" field of package.json'
    ```

    "@prisma/client" will call `require(".prisma")`, so this is needed to do PnP properly.

2.  Properly resolve the location of our virtual dependency ".prisma" in "package.json".

    ```json
    {
	  "resolutions": {
	    ".prisma": "link:node_modules/.prisma"
	  }
    }
    ```

    If you changed the location of generated client-side codes, use your new location in here.

3.  DONE.

### Why `'.prisma': 'link:See "resolution" field of package.json'` ?
We can't do `".prisma": "link:node_modules/.prisma"` in `.yarnrc.yml` since that'll be resulted into `project/.yarn/cache/@prisma-client-npm-2.16.0-d2a6238a45-895e15e1ca.zip/node_modules/@prisma/client/node_modules/.prisma`. We can't do `'.prisma': 'link:../../../../../../node_modules/.prisma'` either since that results into `project/.yarn/cache/@prisma-client-npm-2.16.0-d2a6238a45-895e15e1ca.zip/node_modules/.prisma/client`.

So see need to add a dummy dependency first and fix it at "package.json" later.

**Comment by zachasme** [comment]
Here is an example of how we are using Prisma with Yarn 2 PnP and workspaces/monorepo: [https://github.com/zachasme/prisma-examples/tree/latest/yarn-berry](https://github.com/zachasme/prisma-examples/tree/latest/yarn-berry).

1. Add a new workspace for your prisma client, eg `packages/prisma` with a basic `package.json`:
   ```
   {
     "name": "@example/prisma",
     "main": "index.js",
     "types": "index.d.ts",
     "devDependencies": {
       "@prisma/client": "*",
       "@types/node": "*"
     }
   }
   ```
   also add everything except `package.json` to gitignore. (Note that we need to add `@prisma/client` as a dependency due to #3639 )

2. Make sure the client is generated into that workspace by updating your `schema.prisma`:   
   ```
   generator client {
       provider = "prisma-client-js"
       output   = "../packages/prisma"
   }
   // ...
   ```

3. Make sure to add the workspace to your root `package.json`:
   ```
   "workspaces": [
      "packages/*"
   ],
   ```

4. Then simply import it as `@example/prisma` from any other workspace.

**Comment by yovanoc** [comment]
is there a way without workspaces?

**Comment by Ghirigoro** [comment]
@zachasme That setup helped a lot, but how do you handle migrations? Whatever workspace use the `prisma` workspace will end up with the `prisma.schema` file tucked away in a zip file deep in the bowels of the yarn cache. How do you get prisma migrate to see that file.

Or am I totally off-base? (I just started using Prisma so that's quite likely)

**Comment by zachasme** [comment]
@Ghirigoro Hm, I don't think I had to do anything special. With a structure like [this](https://github.com/zachasme/prisma-examples/tree/latest/yarn-berry):
```
packages/app
packages/prisma      <-- the generated client
prisma/migrations    <-- migrations
prisma/schema.prisma <-- the real schema
```
I can run `yarn prisma migrate status` in the root workspace and it works as expected.

**Comment by darioblanco** [comment]
Hey, another happy user of Yarn2 PnP here (in strict mode, with zero installs, in a monorepo). Or almost quite happy because from time to time I encounter challenges like in this awesome library, which would be perfect if that support comes out of the box.

In any case, I did not like @zachasme idea, because as we have a monorepo and microservices with clear separation in each DB, it did not make sense in our case to have a commonplace for the schemas. However, @zachasme 's idea is awesome which inspired my approach. @simnalamburt was also helpful because it made it easier for me to understand how `prisma` client generation works under Yarn. Thank you guys!

My approach is based also on the `workspace:` protocol (https://yarnpkg.com/features/protocols). It simplifies referencing private packages and as soon as you use a PnP tool for building (like webpack5), you will be fine.

For instance, I separate the backend (e.g. `rest-myapp`) from the prisma schema (e.g. `prisma-myapp`). I usually use prefixes to denote what kind of package it is. 

The root structure:
```
package.json 
packages/rest-myapp
packages/prisma-myapp
```

The `packages/prisma-myapp` structure
```
prisma/schema.prisma
.env # With the DB connection for the dev environment for instance
package.json
```

In `packages/prisma-myapp/package.json`, it is important to define `main` and `types`.  I have only posted the essential scripts (I would recommend to put migration relevant scripts here as well). The `build` will be called by the app package so we will always have the latest client made from our schema. I also needed the `npx prisma` hack to make it work.
```json
{
  "name": "prisma-myapp",
  "private": true,
  "version": "0.0.0",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "yarn prisma generate",
    "prisma": "npx prisma"
  },
  "devDependencies": {
    "@prisma/client": "^2.26.0"
  }
}
```

In `packages/prisma-myapp/prisma/schema.prisma`, do not forget to generate the `dist` folder. I actually put that one in `.gitignore`, because it will always be built whenever you write `yarn` anyway.
```
...
generator client {
  provider = "prisma-client-js"
  output   = "../dist"
}
...
```

Therefore, `packages/rest-myapp/package.json` would look like this:
```json
{
  "name": "rest-myapp",
  "private": true,
  "version": "0.0.0",
  "scripts": {
    "preinstall": "yarn workspace prisma-myapp build",
  },
  "dependencies": {
    "prisma-myapp": "workspace:*"
  }
}
```

When you run `yarn` (or `yarn install`), it will install all the dependencies it find. As `rest-myapp` has a `preinstall` script, it will go to `prisma-myapp` and generate the client in the `dist` folder (which is `.gitignored`, you can push it to your repo if you want to avoid this step, but I guess that is not a good idea). 

Therefore, when you import `prisma-myapp` in the code, it will have all the types from your schema.
```ts
import { PrismaClient } from 'prisma-myapp';
```

The caveat is that you need a 1:1 relationship between backends that use prisma and the prisma schema, but I really hope the community will help to make setting this library with Yarn2 better! If you want to have only one shared workspace, you can still use this approach of course.

**Comment by orbiteleven** [comment]
Interesting note on @zachasme's solution: This works great up until `v2.25.0`, but for `v2.26.0` and `v2.27.0` I get the following: 
```
 ~/W/prisma-examples   latest *…  yarn-berry  yarn prisma generate                                                                                                                       7.5s  Tue Jul 13 15:13:07 2021
Prisma schema loaded from prisma/schema.prisma
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: Done in 0s 180ms
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: Done in 0s 199ms
Error: Could not resolve @prisma/client despite the installation that we just tried.
Please try to install it by hand with npm install @prisma/client and rerun yarn prisma generate 🙏.
```
@zachasme is this the regression to which you're referring in https://github.com/prisma/prisma/issues/3639#issuecomment-756851975?

**Comment by zachasme** [comment]
@orbiteleven it might be related, but I haven't encountered that error. We switched back to the `node-modules` nodeLinker until libraries like prisma gain better support. :-)

**Comment by darioblanco** [comment]
@orbiteleven I use 2.27.0 without problems with my approach and I have strict mode

**Comment by HairyRabbit** [comment]
Any update? looks like 0.28 still can not resolve `@prisma/client`

**Comment by gex** [comment]
> Here's how to make "@prisma/client" work properly with Yarn 2 PnP. If you have a simpler solution, please share!
> 
> 1. Make ".prisma" as a new dependency of "@prisma/client" package.
> 2. Properly resolve the location of our virtual dependency ".prisma" in "package.json".
> 3. DONE.

i try to use your solution because it looks so much easier than juggling with packages (and running npm) when i don't have a monorepo but it doesn't work for me. i'm on `2.29.0` (both `prisma` and `@prisma/client`. i also tried to use a custom output path but [it doesn't work either](https://github.com/prisma/prisma/issues/7234#issuecomment-897370068).

**Comment by simnalamburt** [comment]
@gex Actually… it stopped working even for me since particular version of prisma and I just moved on to the pnpm… Ha ha..

**Comment by gex** [comment]
@simnalamburt thanks for the quick reply! i started to think it's just me but at least i can move on, reverting back the whole project from pnp...

**Comment by kwnath** [comment]
I might be completely wrong and off base here but would installing `@yarnpkg/pnpify` and running it with `yarn pnpify prisma` solve this potentially?

Or are there some downsides I'm unaware of.

**Comment by Globolobo** [comment]
@kwnath, That worked for us, but only on specifically `v2.27`.

To expand on what ☝️ is saying, for those that don't know, you could "[unplug](https://yarnpkg.com/cli/unplug)" your module, and then use [pnpify](https://yarnpkg.com/advanced/pnpify) to run as pnp'd module.
 We did this, within our workspace: 
 
`yarn unplug prisma @prisma/client`

 and then run the prisma commands
 
`yarn pnpify prisma generate`

**Comment by elpddev** [comment]
Currently on our POC, `yarn prisma generate` works. To use the generated packages:
1. we copy `.prisma/client` generated package from the `node_modules` folder to the `packages` workspaces folder 
2. Rename the package to `@my-company/prisma-client`. 
3. We had to also specify in the `package.json` of the generated package that it depended on `@prisma/client`. 

After that, in any other package in the monorepo we can do `import {PrismaClient} from '@my-company/prisma-client'` and use it to query the db.

**Comment by dgattey** [comment]
Thank you @Globolobo + @darioblanco for putting me on the right path! I compiled this year-of-2022 note since no one suggestion above worked perfectly for me with Next + Vercel + Yarn 3 + Yarn PnP + Yarn Zero Install. Here's my setup.

---

In `package.json` I use pnpify with a script to wrap generation. I also run generation as part of my `yarn build`. Nothing fancy with module resolution needed.
```json
"scripts": {
    "build": "yarn db:generate && next build",
    "db:generate": "yarn pnpify prisma generate",
},
"dependencies": {
    "@prisma/client": "3.8.1",
},
"devDependencies": {
    "@yarnpkg/pnpify": "^3.1.1-rc.12",
    "prisma": "3.8.1",
  },
```
In my `prisma/schema.prisma`, I set a relative output path (this can be anything, I'm just putting it into my server api folder). It just needs to be within `src`.
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../src/api/server/generated"
}

datasource db {
  provider = "mysql"
  url = env("DATABASE_URL")
}
```

I make sure to ignore that path for version control (and prettier/eslint/anything else) in `.gitignore`/etc
```gitignore
# Generated files from Prisma
src/api/server/generated
```

And finally, in my `.ts` files, I just import from my custom path instead of from the main package. Works great!
```typescript
import { PrismaClient } from 'api/server/generated/';
const dbClient = new PrismaClient();
```
or 
```typescript
import type { Prisma } from 'api/server/generated/'';
```
---

### Notes

1. If your package resolution requires a relative path like `../../api/...` or a prefix like `@app/api/...`, that should work fine instead, I just use `api/...` directly since I set that up as a top level path.

2. You could avoid putting the generate command in your build command if you check the generated folder into version control, but I didn't think that was worth it. If you don't check it in, then Vercel/whatever will fail the build since it'll be missing until you run generation again. As my DB should always be up and accessible, I see no problem with generating from Vercel before I build. Catches breaking changes anyway.

**Comment by behrangsa** [comment]
This is still broken.

Inside a new empty folder, run:

```
$ yarn init -2
```

Then follow the steps [here](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases-typescript-postgres), but with yarn instead of npm:

```
$ yarn add prisma typescript ts-node @types/node --dev
```

Then run prisma:

```
$ yarn exec prisma
```

So for so good.

However, when I add the models and then run:

```
$ yarn exec prisma migrate dev --name init
```

Everything breaks:

```
PS C:\Users\Me\Documents\Projects\local\yarn-prisma-bug> yarn exec prisma migrate dev --name init
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "mydb", schema "public" at "localhost:49153"

PostgreSQL database mydb created at localhost:49153

Applying migration `20220720071915_init`

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20220720071915_init/
    └─ migration.sql

Your database is now in sync with your schema.

Running generate... (Use --skip-generate to skip the generators)
Error: Could not resolve prisma and @prisma/client in the current project. Please install them with yarn add -D prisma and yarn add @prisma/client, and rerun yarn prisma generate 🙏.
```

The hint _"Could not resolve prisma and @prisma/client in the current project. Please install them with yarn add -D prisma and yarn add @prisma/client, and rerun yarn prisma generate 🙏."_ doesn't help.

**Comment by ivangabriele** [comment]
@dgattey [https://github.com/prisma/prisma/issues/1439#issuecomment-1023884266](solution) works well.

The only issue left is that prisma automatically calls generate when we run a `migrate` in prod or dev mode which throws an error since `prisma generate` must be called via `pnpify`.

**Comment by janpio** [comment]
@dgattey Just to clarify: Do you get the same error when you run `yarn exec prisma generate` instead of the `migrate dev` command? That would help to isolate exactly what code is causing problems here. (Hunch: The auto installation of Prisma CLI and Prisma Client in the Client generation is not compatible with Yarn 2.)

**Comment by dgattey** [comment]
@janpio i think you mean to ask that of @behrangsa? My post was about making generate work.

**Comment by jdpowell1** [comment]
@janpio with our setup, we're getting the same error on both generation and migration. We have to use pnpify when generating and we have to skip the generation on migration and use pnpify instead.

**Comment by janpio** [comment]
Ok, that makes sense - `migrate dev` tries to run the command that can not work for your setup.

**Comment by pencilcheck** [comment]
wonder if there is a way to not use npm?? it keeps generating package-lock and i'm using yarn for the project

**Comment by behrangsa** [comment]
@janpio not sure. I will check and circle back to you.

**Comment by DanielHoffmann** [comment]
I just want to drop by and say what worked for me, I am using nexus-prisma so I am posting that setup here as well:

set "output" in `schema.prisma` generators to a folder within your project (add it to gitignore):

```
generator client {
  provider = "prisma-client-js"
  output   = "./.prisma"
}

generator nexusPrisma {
  provider = "nexus-prisma"
  output   = "./.nexus-prisma"
}
```

add `@yarnpkg/pnpify`: 
```
yarn add @yarnpkg/pnpify
```

add this to your `package.json`:
```json
"scripts": {
    "db-generate": "rm -rf prisma/.prisma && rm -rf prisma/.prisma-nexus && yarn pnpify prisma generate",
    "db-push": "yarn pnpify prisma db push",
    "db-migrate-reset": "yarn pnpify prisma migrate reset",
    "db-migrate-dev": "yarn pnpify prisma migrate dev"
},
"dependencies": {
    ".nexus-prisma": "link:./prisma/.nexus-prisma/",
    ".prisma": "link:./prisma/.prisma/",
}
```

prefix any other prisma CLI commands you use with `yarn pnpify` as well


run `yarn install` and `yarn run db-generate`


The only setting I have in my `.yarnrc.yml` is `pnpMode: strict`. VSCode Intellisense can also pick up changes from `db-generate` just fine

**Comment by isaachinman** [comment]
@janpio Can we please have an official response on any timeline for this issue?

**Comment by janpio** [comment]
We do no communicate any timelines. We are aware the issue exists, we acknowledge it is a real problem that makes it harder to use Prisma, and when we have capacity to work on issues like this, we will. (If this ever changes, we will communicate it by explaining this here in the issue and close it.) Sorry I can't give you a more satisfying answer.

**Comment by iNViTiON** [comment]
I've managed to build a Docker image for an application using Yarn PnP. This is how I got it to work.
I'm using Prisma 5 in monorepo on Nx 16.7 with Yarn 4. Because of monorepo structure I using this Prisma client configuration & path.

```Prisma
generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native"]
  engineType      = "binary"
  output          = "../../../node_modules/@prisma/client/demo-api"
}
```
Then I import the client with the following line.
```ts
import { PrismaClient } from '@prisma/client/demo-api';
```

In the `Dockerfile` I use to build the application image, I've included additional Yarn configuration.
```Dockerfile
RUN yarn config set enableGlobalCache false && \
  yarn config set enableScripts false && \
  yarn config set pnpMode loose && \
  npm pkg set --json dependenciesMeta.@prisma/client.built=true \
  dependenciesMeta.@prisma/client.unplugged=true \
  dependenciesMeta.@prisma/engines.unplugged=true \
  dependenciesMeta.prisma.unplugged=true
```

Next, I modify the new node_modules path to match the environment outside the monorepo.
```Dockerfile
RUN sed -i 's|../../../node_modules|./node_modules|g' schema.prisma
```

Finally, I generate the Prisma client and map the specific path for PnP and doing the cleanup.
```Dockerfile
RUN yarn pnpify prisma generate && \
  npm pkg set dependencies[".prisma"]=link:$(ls -d .yarn/unplugged/@prisma-client*)/node_modules/.prisma && \
  yarn remove @yarnpkg/pnpify prisma
```

This was a bit tricky, but it works smoothly now.

**Comment by seanparkross** [comment]
+1
:(

**Comment by theoparis** [comment]
Unfortunately, this also caused an issue when I tried to setup https://github.com/linkwarden/linkwarden, which uses prisma 😕  I'm forced to use pnpm because of this issue (npm is too slow and uses too much disk space).

**Comment by zach-betz-hln** [comment]
Related: https://github.com/prisma/prisma/issues/8765#issuecomment-1854451597

**Comment by kopax** [comment]
Hello, I am upgrading from yarn v1 to yarn v4, and I am hitting the issue. 

I am amazed that it is open since 4 years already. This should be in the documentation. 

I had to upgrade lerna as linking in lerna project was causing a lot of hassle, also during docker build images. 


Does anybody have a workaround to propose ?

**Comment by kopax** [comment]
That's a bit of a nightmare.

I had to upgrade yarn to v4, and since my prisma generate client stopped to work.

I have found a way in this thread to generate it using pnipfy and using my source folder, however, my project also use typescript and two files have typescript compile error that can't be ignored with `compileOptions.checkJs: false`.

The two errors are (just a sample):


```
src/generated/client/index.js:2:1 - error TS9006: Declaration emit for this file requires using private name 'TracingHelper' from module '"/Users/dka/workspace/github.com/kopax/montagne2/apps/api/src/generated/client/runtime/library"'. An explicit type annotation may unblock declaration emit.

2 Object.defineProperty(exports, "__esModule", { value: true });
  ~~~~~~

src/generated/client/runtime/edge-esm.js:1:1 - error TS9005: Declaration emit for this file requires using private name 'i'. An explicit type annotation may unblock declaration emit.
```


Found 15 errors in 2 files.

Errors  Files
    14  src/generated/client/index.js:2
     1  src/generated/client/runtime/edge-esm.js:1

It is worst when I set `checkJs: false`.

Currently, there is absolutely no way to disable those compile error and thus, I can't use that client if I keep yarn 4.

**Comment by zach-betz-hln** [comment]
@kopax does using the [binary engine type](https://github.com/prisma/prisma/issues/8765#issuecomment-1854451597) make a difference ?

**Comment by renatoaraujoc** [comment]
This situation everyone is going thru is easy to solve.

First, **accept the fact from Prisma team** for now that non node-modules folders **will break the installation of prisma**.

**Solution:**

When running `yarn install` (v4 in my case), you must have `.yarnrc.yml` file at the same level of your `package.json`, it seems like yarn install doesn't provide an option like "--linker=node-modules", so you need this file, contents:

```yml
nodeLinker: node-modules
```

Then run your install and everything will work, results below for my Dockerfile build:

![Screenshot 2024-09-25 at 19 23 25](https://github.com/user-attachments/assets/013452ee-07b8-4ae6-8012-67ef3478e6c8)

**Comment by zach-betz-hln** [comment]
Hi @renatoaraujoc - if you prefer to keep `nodeLinker: pnp` then you can use a combination of these solutions: 
- [part 1](https://github.com/prisma/prisma/issues/8765#issuecomment-1086611596) 
- [part 2](https://github.com/prisma/prisma/issues/8765#issuecomment-1854451597)

**Comment by renatoaraujoc** [comment]
Hey @zach-betz-hln,

That solution also works but (for me) it's just easier to throw the client in node_modules so you can just import from @prisma/client. Its true that you can just make some sort of TSConfig pathMapping and exclude it in the .gitignore, but I'm too lazy for this :D

Any way, both solutions work at least.

**Comment by headhunter45** [comment]
For anyone still struggling with this. I created a minimal repo with some documentation about specifically how to get Prisma to work in a Yarn 4 PnP environment by sticking it in its own workspace. Instead of throwing random code in your config you can check it out and see why and how it works. The other approach I tried was doing all of the Prisma generation in a docker container. It worked, but I wanted to remove the need to run another container during a build. 

https://github.com/headhunter45/YarnPrismaPnP

**Comment by andrey-debbie** [comment]
wtf guys, it's time to fix this

**Comment by andrey-debbie** [comment]
```
yarn add @yarnpkg/pnpify
yarn pnpify prisma generate
```

```
import { PrismaClient } from '../<path_to>/generated/prisma'
```
